### ISSUE TITLE
Концептуальная схема

### ISSUE THREAD

**Issue by chaos23d** [BMSTU2024/-#5]

**Comment by chaos23d** [comment]
Я прикрепил концептуальную схему. Кое-какие моменты поправил.

Описание предметной области:

-К каждому человеку прикреплён 1 пользователь(далее просто игрок). Игрок содержит логин, пароль и время прохождения
-Достижение - некоторая ачивка, которая появляется при исполнение некоторого условия. У каждого игрока собственная информация о достижениях
-Информация о конкретном достижении закреплена на 1 игрока
-Уровни есть в режиме кампании(одиночная игра). У каждого игрока имеется собственная информация об уровнях. (К примеру у 1 игрока 3 уровень пройден,когда у 2 нет ).
-Информация о конкретном уровне закреплена на 1 игрока
-Игрок может являться участником клана. Игрок может находиться только в 1 клане
-В клане содержаться множество участников с собственным статусом
-Чат это группа игроков, где они переписываются. Она имеет название
-Игрок может находиться в нескольких чатах; в чате множество игроков. информация, кто в каком чате находится , собирается в списке участников чата
-Список сообщений содержит отправителя конкретного чата и сам текст сообщения
![КС](https://github.com/user-attachments/assets/73febf1e-7ffb-4bf8-810f-d4cf9b061ab6)

**Comment by chaos23d** [comment]
Насчёт 4 сущностей и их связей(чат, список участников чата, список сообщений и игрок). 

Просто на созвоне я как-то не так рассказал, или плохо объяснил (мне говорят, что порой плохо объясняю). Поэтому попробую на примере

Есть 4 игроков и 2 чата (В данном случае будут интересовать только логины игроков и название чатов). Отправитель сообщения 1, но получатели все участники чата(кроме отправителя), в котором написано сообщение

Игроки [kirill, kost] в чате [dinamit]; Игроки [kirill, rock55, Demon] в чате [bmstu].

Игрок [kirill] из чата [dinamit] написал сообщение [Сегодня будем играть?];Игрок [kost] из чата [dinamit] написал сообщение [не, не получится?];Игрок [Demon] из чата [bmstu] написал сообщение [Кто-то знает, как лучше всего пройти 3 уровень?];Игрок [kirill] из чата [bmstu] написал сообщение [я знаю, но сейчас не отвечу]

На рисунке есть 2 таблицы. нижняя таблица связана с верхней (столбец отправитель(будет считаться внешним ключом) из нижней таблицы связан со столбцом id(будет считаться первичным ключом) в верхней таблице )

Надеюсь вы меня правильно поймёте

![1](https://github.com/user-attachments/assets/8c6db68d-8dac-4f5c-8795-be97f6736e11)

**Comment by chaos23d** [comment]
Да, вы правы. Стрелки надо местами поменять. Просто в сущности Список участников чата атрибуты [чат, игрок] будут внешними ключами, поэтому для 1 чата(или игрока) будет множество объектов из сущности Список участников чата. До этого особо не доводилось детально реализовывать связь многие ко многим

**Comment by chaos23d** [comment]
![КС](https://github.com/user-attachments/assets/fcdabe48-6bbd-4579-a7a8-afde78305a3c)

Я переделал так как надо.

Описание предметной области:

-К каждому человеку прикреплён 1 пользователь(далее просто игрок). Игрок содержит логин, пароль и время прохождения уровней(время кампании)

-Достижение - некоторая ачивка, которая появляется при исполнение некоторого условия.
-Каждое достижение может принадлежать всем игрокам 
-Игрок может иметь несколько достижений
-Списки достижений связывает игроков и достижения с индивидуальным статусом(к примеру есть 2 игрока и 2 достижения, и у каждого игрока собственный статус конкретного достижения). Статус -числовой параметр (некое количество)

-Уровни есть в режиме кампании(одиночная игра) и они пронумерованы.
-Каждый уровень может принадлежать всем игрокам 
-Игрок может иметь несколько уровней
-Список уровней связывает игроков и уровни с индивидуальным количеством звёзд, которые показывают , как пройден уровень(к примеру есть 2 игрока и 2 уровня, и у каждого игрока собственное количество звёзд уровня). Кол-во звёзд от 0 до 3(если звёзд больше 0, то уровень пройден)

-Игрок может являться, может не являться участником клана. Игрок может находиться только в 1 клане с собственным статусом. В клане несколько статусов: участник, замглава, глава(может быть только 1)
-В клане содержаться множество участников

-Чат это группа игроков, где они переписываются. Она имеет название
-Игрок может находиться в нескольких чатах
-В чате множество игроков
-Список участников чатов связывает игроков и чаты с индивидуальным статусом. Статус может принимать несколько значений: участник, модератор, администратор(может быть только 1)

-Сообщение - некое письмо , написанное всем участникам чата(чата, в котором находится отправитель) в определённое время
-конкретный участник чата может отправлять множество сообщений
-У сообщения только 1 отправитель

**Comment by wldmr-key** [comment]
по существу уже гораздо лучше, но

Страшный грех = дублирование связей атрибутами. В данном случае стрелка от игроков к спискам участников чатов подразумевает связь игрока (и, следовательно, атрибут “игрок” включать нельзя), а от списка участников чатов к спискам сообщений подразумевает связь по имени игрока и чату и, следовательно, в сущности “списки сообщений! не должно быть атрибутов “игрок” (здесь он, к тому же еще и назван “пользователь, видимо, чтобы всех запутать) и “чат”. Это не единственный пример - так у тебя почти везде

Крайне неудачные названия сущностей: какие “списки”? Разве списков много? Называй, уж, или “список участников чатов” или ”участники чатов”!

**Comment by chaos23d** [comment]
![в](https://github.com/user-attachments/assets/80e5acba-bea3-41ba-ab40-d7ca2d73a2d0)
Решил правильно оформить КС. Удалил некоторые лишние сущности. Она соответствует также логической схеме. 

Кстати, есть вопрос. Я добавил такие сущности, связанные с достижениями, уровнями и кланами. В проекте у меня только реализована часть с чатами. Ничего, если не успею реализовать? Просто дальше хотел доделать управление с чатами(добавление, удаление участников), а дальше уже с кланами и достижениями. Просто времени для уровней может не хватить на реализацию(это по большей части касается с Режимом уровней и настройки каждого уровня, нежели с обращением к БД)

**Comment by chaos23d** [comment]
![КС](https://github.com/user-attachments/assets/9eafbbb1-da21-445a-8062-91db02ef81da)

Я немного исправил КС под нынешнее ТЗ

Описание предметной области:

-К каждому человеку прикреплён 1 пользователь(далее просто игрок). Игрок содержит логин, пароль и статус нахождения в сети(в сети или не в сети)

-Чат это группа игроков, где они переписываются. Она имеет название
-Игрок может находиться в нескольких чатах
-В чате множество игроков
-Список участников чатов связывает игроков и чаты с индивидуальным статусом. Статус может принимать несколько значений: участник, администратор(может быть только 1),ожидающий (приглашён, но ещё не в чате), удалённый(тот, кто когда-то был в чате)

-Сообщение - некое письмо , написанное всем участникам чата(чата, в котором находится отправитель) в определённое время
-конкретный участник чата может отправлять множество сообщений
-У сообщения только 1 отправитель

-Сражение это игра между 2 игроками. Может иметь номер, начало и окончание

-История сражений показывает, какие игроки закончили сражение с каким результатом. Результат может быть выигрыш или проигрыш
-Один игрок мог участвовать во многих сражениях
-В сражении могут участвовать множество игроков
